File: leetcode/131.palindrome-partitioning.py
#
# @lc app=leetcode id=131 lang=python3
#
# [131] Palindrome Partitioning
#

# @lc code=start
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def isPalindrome(s):
            if s == s[::-1]:

                return True
            else:
                return False

        res = []
        part = []

        def dfs(i):
            if i >= len(s):
                res.append(part[:])
                return
            for j in range(i, len(s)):
                if isPalindrome(s[i : j + 1]):
                    part.append(s[i : j + 1])
                    dfs(j + 1)
                    part.pop()

        dfs(0)
        return res


# @lc code=end

File: leetcode/131.palindrome-partitioning.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode id=131 lang=java
 *
 * [131] Palindrome Partitioning
 */

// @lc code=start
class Solution {
    List<List<String>> res = new ArrayList<>();
    List<String> part = new ArrayList<>();
    String s;

    public List<List<String>> partition(String s) {
        this.s = s;
        dfs(0);
        return res;
    }

    public boolean isPalindrome(String s, int start, int end) { // A simple palindromic function start from 0 go till
                                                                // end. And basically keep on checking till they don't
                                                                // cross.
        while (start <= end) {
            if (s.charAt(start++) != s.charAt(end--))
                return false;
        }
        return true;
    }

    public void dfs(int i) {
        if (i >= s.length()) {
            res.add(new ArrayList<>(part));
        }
        for (int j = i; j < s.length(); j++) {
            if (isPalindrome(s, i, j)) {
                part.add(s.substring(i, j + 1));
                dfs(j + 1);
                part.remove(part.size() - 1);
            }
        }
    }
}
// @lc code=end
